### PR TITLE
[PyTorch] _addm_activation native function for matmul/bias/activation fusion

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #75493
* #75347
* #75346
* #75491
* #75370
* #74629
* __->__ #74490

Here's an extended version of addmm that takes advantage of cublasLt's fused addmm + relu/gelu support.

Differential Revision: [D35019612](https://our.internmc.facebook.com/intern/diff/D35019612/)

### CODE REVIEW

**Comment by ngimel** [comment]
cc @xwang233

File: aten/src/ATen/native/cuda/Blas.cpp [hunk 1]
```diff
@@ -263,6 +264,15 @@ Tensor& addmm_out_cuda_impl(Tensor& result, const Tensor& self, const Tensor& ma
               result_ptr,
               result_ld);
         });
+    switch (activation) {
+      case cuda::blas::GEMMAndBiasActivationEpilogue::RELU:
+        at::relu_(const_cast<Tensor&>(*result_));
+        break;
+      case cuda::blas::GEMMAndBiasActivationEpilogue::GELU:
+        result_ = c10::MaybeOwned<Tensor>::owned(at::gelu(*result_));
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Blas.cpp:272]: why relu_ is inplace, but gelu isn't?

**Review by ngimel — COMMENTED** [review]
This approach looks fine. We won't have backward for those functions, but for private functions that's ok. Do we want to expose them in native_functions.yaml if they are private and intended to be called only from some other publicly exposed functions?

File: aten/src/ATen/native/cuda/Blas.cpp (hunk 1 above)
Comment (swolchok) [reply aten/src/ATen/native/cuda/Blas.cpp:272]: IIRC we don't have in-place gelu, but maybe I did something wrong.

Comment (swolchok) [reply aten/src/ATen/native/cuda/Blas.cpp:272]: Looks like we have gelu.out and regular gelu, but no in-place.

**Comment by swolchok** [comment]
> Do we want to expose them in native_functions.yaml if they are private and intended to be called only from some other publicly exposed functions?

Don't we need to do that for testing?

Also, do we want to keep this private? Fused matmul-bias-{r,g}elu  seems like something people might want (but what I do I know about AI research).

**Comment by ngimel** [comment]
Yeah, we need it to be able to test from python, and yes, fused matmul-bias-gelu/relu is a useful op, but requirements are higher for publicly available op, such as you are expected to provide a gradient computation for it, and documentation. We still can have them in native_functions as private ops starting with `_`, we have a few of those already.

**Comment by swolchok** [comment]
hey @ngimel this is looking green, can I have a review?

File: aten/src/ATen/native/cuda/Blas.cpp [hunk 2]
```diff
@@ -263,8 +299,23 @@ Tensor& addmm_out_cuda_impl(Tensor& result, const Tensor& self, const Tensor& ma
...
+        break;
+      case Activation::GELU:
+        result_ = c10::MaybeOwned<Tensor>::owned(at::gelu(*result_));
+        break;
+      default: break;
+    }
   }
 
+#if defined(CUDA_VERSION) && CUDA_VERSION >= 11000 && CUDA_VERSION < 11060 && !defined(_MSC_VER)
+  if (useLtInterface && activation == Activation::GELU) {
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Blas.cpp:314]: can you also leave a comment here explaining why gelu has to be done separately

**Review by ngimel — APPROVED** [review]

**Comment by ngimel** [comment]
cc @ptrblck. @swolchok, @ptrblck says he couldn't reproduce faillure with gelu activation with the sizes that you provided, do you have a test that would reproduce it?

File: aten/src/ATen/native/cuda/Blas.cpp [hunk 3]
```diff
@@ -231,7 +256,18 @@ Tensor& addmm_out_cuda_impl(Tensor& result, const Tensor& self, const Tensor& ma
               mat2_ld,
               self.data_ptr<scalar_t>(),
               result_->data_ptr<scalar_t>(),
-              result_ld);
+              result_ld,
+#if CUDA_VERSION >= 11060
+              activation_to_gemm_and_blas_arg(activation)
```
[the remark this answers is not in the document]
Comment (swolchok) [reply aten/src/ATen/native/cuda/Blas.cpp:261]: just to pacify the tests, commenting it out is fine. don't know how we will know when it is OK though

File: aten/src/ATen/native/cuda/Blas.cpp (hunk 2 above)
Comment (swolchok) [reply aten/src/ATen/native/cuda/Blas.cpp:314]: OK I'll make the above gelu comment into a NOTE and reference it

**Comment by swolchok** [comment]
> cc @ptrblck. @swolchok, @ptrblck says he couldn't reproduce faillure with gelu activation with the sizes that you provided, do you have a test that would reproduce it?

I'll see what I can do.

**Comment by swolchok** [comment]
> cc @ptrblck. @swolchok, @ptrblck says he couldn't reproduce faillure with gelu activation with the sizes that you provided, do you have a test that would reproduce it?

I have included a repro in test_linalg.py (which I am happy to remove before checkin since it's huge). It should fail on CUDA 11.4 if the `#if 0` and `#if !0` in Blas.cpp are flipped (to `#if !0` and `#if 0` respectively).

**Comment by swolchok** [comment]
> > cc @ptrblck. @swolchok, @ptrblck says he couldn't reproduce faillure with gelu activation with the sizes that you provided, do you have a test that would reproduce it?
> 
> I have included a repro in test_linalg.py (which I am happy to remove before checkin since it's huge). It should fail on CUDA 11.4 if the `#if 0` and `#if !0` in Blas.cpp are flipped (to `#if !0` and `#if 0` respectively).

Needs a further update before you will be able to run it, but I have to run. I will fix it up tomorrow morning.

**Comment by swolchok** [comment]
> > > cc @ptrblck. @swolchok, @ptrblck says he couldn't reproduce faillure with gelu activation with the sizes that you provided, do you have a test that would reproduce it?
> > 
> > 
> > I have included a repro in test_linalg.py (which I am happy to remove before checkin since it's huge). It should fail on CUDA 11.4 if the `#if 0` and `#if !0` in Blas.cpp are flipped (to `#if !0` and `#if 0` respectively).
> 
> Needs a further update before you will be able to run it, but I have to run. I will fix it up tomorrow morning.

OK I've verified that `test_addmm_repro_possible_cublas_gelu_bug_cuda_float32` (and bfloat16, and float64) will fail as advertised if and only if the following patch is applied:

```
--- a/fbcode/caffe2/aten/src/ATen/native/cuda/Blas.cpp
+++ b/fbcode/caffe2/aten/src/ATen/native/cuda/Blas.cpp
@@ -257,7 +257,7 @@
               self.data_ptr<scalar_t>(),
               result_->data_ptr<scalar_t>(),
               result_ld,
-#if 0
+#if !0
               activation_to_gemm_and_blas_arg(activation)
 #else
               // GELU is not supported (and does not compile!) prior
@@ -315,7 +315,7 @@
 // gating activation_to_gemm_and_blas_arg above; here we are manually
 // performing a post-GELU because we weren't able to use the GELU
 // epilogue above.
-#if !0
+#if !!0
   if (useLtInterface && activation == Activation::GELU) {
     result_ = c10::MaybeOwned<Tensor>::owned(at::gelu(*result_));
   }
```

That said, I agree it is suspicious that it doesn't fail when you try via cublas and it does fail when I try via my new codepath, so I am going to look through this diff again to make sure I haven't done something silly like double-gelu.

**Comment by swolchok** [comment]
> I am going to look through this diff again to make sure I haven't done something silly like double-gelu.

Nope, didn't find anything silly. We know relu is working, and we're functioning correctly for gelu and set up to do the fusion when we figure out correctness, so I'm going to go ahead and land this.

**Comment by swolchok** [comment]
https://gist.github.com/swolchok/f550a9dfbab879cc514912d5f5ab59af has the version of test_linalg.py with the repro test for gelu epilogue not working properly, in case old versions of this PR aren't accessible after I update it.

**Comment by eqy** [comment]
Thanks for the repro; I've isolated it to cublaslt and have passed the repro on.

**Pull request closed without merging**